app/validator/element.ts

```diff
@@ -1,71 +1,71 @@
+        this.client = createClient({
             baseUrl: homeServerUrl,
             accessToken: accessToken,
             userId: userId,
-            cacheLevel: -1,
         });
-        this.chainName = 'kusama';
-        this.maxRetries = 10;
-        this.pollingRoomMessageInterval = 3;
-        this.pollingRoomInterval = 2;
     }
 
-    async _invoke(riotAccount, token, account) {
-        let self = this;
+    async _invoke(riotAccount: string, token: string, account: string) {
         let roomId = null;
-        let client = self.client;
 
         /// If the client is not running, start the client
-        if (!client.clientRunning) {
+        // @ts-ignore
+        if (!this.client.clientRunning) {
             /// NOTE: Do nothing, but listening for incoming messages. Do not remove !
             /// Capture user input events
-            client.on('Room.timeline', async function (/* event, room, toStartOfTimeline, removed, data */) {});
-            await client.startClient();
+            this.client.on('Room.timeline', async function (/* event, room, toStartOfTimeline, removed, data */) {});
+            await this.client.startClient();
         }
 
         const results = await RiotCollection.query({ riot: riotAccount });
         if (!_.isEmpty(results) && results[0].roomId) {
             roomId = results[0].roomId;
             logger.debug(`Already has room for riot user: ${riotAccount}, room id: ${roomId}`);
         } else {
-            roomId = await self.createRoom(riotAccount);
+            roomId = await this.createRoom(riotAccount);
         }
         /// Fetch this room from remote.
         /// Make sure we obtain an avaiable room instance
         /// If we cannot fetch the room instance, it means this room is destroyed by user (admin) manually.
         /// We need to create a new one.
-        let room = await self.getRoom(roomId);
+        let room = await this.getRoom(roomId);
         if (_.isEmpty(room)) {
             logger.debug(`Create a new room since the old room ${roomId} isn't found`);
-            roomId = await self.createRoom(riotAccount);
-            room = await self.getRoom(roomId);
+            roomId = await this.createRoom(riotAccount);
+            room = await this.getRoom(roomId);
         }
 
         /// Check if the target riot account join our room or not
         /// If yes, we needn't to invite him/her again
         /// otherwise, we send an invitation request to target riot user
-        const joinedMembers = room.getMembersWithMembership('join');
+        const joinedMembers = room!.getMembersWithMembership('join');
         const isNotJoined = _.isEmpty(_.find(joinedMembers, ['userId', riotAccount]));
 
         if (isNotJoined) {
             logger.debug(`Riot user ${riotAccount} isn't joined the room.`);
-            const invitedMembers = room.getMembersWithMembership('invite');
+            const invitedMembers = room!.getMembersWithMembership('invite');
```

Comment:
what does `!` after `room` mean?

Reply:
It's telling typescript i definitely exists. We need it because of the way this is written:

```typescript
 let room = await this.getRoom(roomId);
        if (_.isEmpty(room)) {
            logger.debug(`Create a new room since the old room ${roomId} isn't found`);
            roomId = await this.createRoom(riotAccount);
            room = await this.getRoom(roomId);
        }
```
where this.getRoom returns `Promise<Room | null>` but we know it's not null as we created he room just beofer